### PR TITLE
New Resource: aws_cloudtrail_event_data_store

### PR DESCRIPTION
**Pull request by bschaatsbergen**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #22434 

Output from acceptance testing:

```
$ make testacc TESTS=TestAccEventDataStore PKG=cloudtrail
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudtrail/... -v -count 1 -parallel 20 -run='TestAccEventDataStore'  -timeout 180m
=== RUN   TestAccEventDataStore_basic
=== PAUSE TestAccEventDataStore_basic
=== RUN   TestAccEventDataStore_disappears
=== PAUSE TestAccEventDataStore_disappears
=== RUN   TestAccEventDataStore_multi_region_enabled
=== PAUSE TestAccEventDataStore_multi_region_enabled
=== RUN   TestAccEventDataStore_advanced_event_selector
=== PAUSE TestAccEventDataStore_advanced_event_selector
=== CONT  TestAccEventDataStore_basic
=== CONT  TestAccEventDataStore_multi_region_enabled
=== CONT  TestAccEventDataStore_advanced_event_selector
=== CONT  TestAccEventDataStore_disappears
--- PASS: TestAccEventDataStore_disappears (17.64s)
--- PASS: TestAccEventDataStore_multi_region_enabled (21.19s)
--- PASS: TestAccEventDataStore_basic (22.35s)
--- PASS: TestAccEventDataStore_advanced_event_selector (24.41s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/cloudtrail 24.473s
```

### CODE REVIEW

**Review by NathanGitgit — COMMENTED** [review]
Great initiative, I'm looking forward to keep track of this draft PR.

**Comment by NathanGitgit** [comment]
👍

File: internal/provider/provider.go [hunk 1]
```diff
@@ -898,7 +898,8 @@ func Provider() *schema.Provider {
 			"aws_cloudhsm_v2_cluster": cloudhsmv2.ResourceCluster(),
 			"aws_cloudhsm_v2_hsm":     cloudhsmv2.ResourceHSM(),
 
-			"aws_cloudtrail": cloudtrail.ResourceCloudTrail(),
+			"aws_cloudtrail":      cloudtrail.ResourceCloudTrail(),
+			"aws_cloudtrail_lake": cloudtrail.ResourceCloudTrailLake(),
```
[the remark this answers is not in the document]
Comment (bschaatsbergen) [reply internal/provider/provider.go:902]: Snap, thanks Kit!

**Comment by bschaatsbergen** [comment]
@ewbankkit could you advice me on how to implement the creation and update of the `AdvancedEventSelector`.

For the read I'm re-using the `flattenAdvancedEventSelector` as following:

`d.Set("advanced_event_selector", flattenAdvancedEventSelector(eventDataStore.AdvancedEventSelectors))` 

But for the `resourceEventDataStoreCreate` and `resourceEventDataStoreUpdate` I'm not quite sure how to tackle this. I was looking at the logic for the AdvancedEventSelectors in `cloudtrail.go` but found myself a bit stuck when trying to use that.

Thanks in advance,

**Comment by bschaatsbergen** [comment]
Hey @ewbankkit,

The event data store CRUD seems to be work, besides the `advanced_event_selector`.

By default the SDK implicitly sets it to 'capture all management events'. Because this is not reflected in the resource when declaring it, it does end up in the state, but the subsequent `plan` or `apply` will again try to delete the `advanced_event_selector` that's set by the default behaviour.

Could you advice me on how to tackle this please?

**Comment by bschaatsbergen** [comment]
Hey @ewbankkit,

The event data store CRUD seems to work, there's just something wrong with the `advanced_event_selector`.

By default the SDK implicitly sets it to 'capture all management events'. Because this is not reflected in the resource when declaring it, it does end up in the state, but the subsequent `plan` or `apply` will again try to delete the `advanced_event_selector` that's set by the default behaviour.

```
  # aws_cloudtrail_event_data_store.main will be updated in-place
  ~ resource "aws_cloudtrail_event_data_store" "main" {
        id                             = "arn:aws:cloudtrail:eu-central-1:123456789010:eventdatastore/4481b52a-4892-4c3e-a2b0-b4ec4e40368d"
        name                           = "tf-test-store-7"
      ~ tags                           = {
          + "Name" = "tf-test-store-7"
        }
      ~ tags_all                       = {
          + "Name" = "tf-test-store-7"
        }
        # (5 unchanged attributes hidden)

      - advanced_event_selector {
          - name = "Default management events" -> null

          - field_selector {
              - ends_with       = [] -> null
              - equals          = [
                  - "Management",
                ] -> null
              - field           = "eventCategory" -> null
              - not_ends_with   = [] -> null
              - not_equals      = [] -> null
              - not_starts_with = [] -> null
              - starts_with     = [] -> null
            }
        }
    }

```
Plan: 0 to add, 1 to change, 0 to destroy.

Could you advice me on how to tackle this please?

**Comment by ewbankkit** [comment]
@bschaatsbergen Try marking `advanced_event_selector` (and its nested attributes) as [`Computed`](https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors#computed) together with `Optional`.

**Comment by bschaatsbergen** [comment]
Thanks @ewbankkit, I'll give that a try. I noticed that if you only set the `field` in the `field_selector` (without providing an operator) the API returns a 400 (which is apparently retryable) and it continues to `Still modifying...`. It should actually print an error as a operator is always required in combination with a `field`.

**Comment by bschaatsbergen** [comment]
Hey @ewbankkit, if you have some spare time would you mind checking out why my `TestAccEventDataStore_disappears` test seems to fail? It does show me some weird behaviour. When I provision the resource through terraform, then delete it from the console, and then run a plan/apply it says that 'nothing has changed' and everything seems ok. (even though it's pending deletion).

**Comment by bschaatsbergen** [comment]
Done @ewbankkit, besides the previous comment I've put on this PR (which seems related to the AWS API) and that I'm unable to test the `organization_enabled`, everything is tested and seems to work :+1:

**Comment by bschaatsbergen** [comment]
Done @ewbankkit, besides the previous comment (which seems related to the AWS API) and that I'm unable to test the organization_enabled, everything is tested and seems to work 👍

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TESTS=TestAccEventDataStore_ PKG=cloudtrail 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudtrail/... -v -count 1 -parallel 20 -run='TestAccEventDataStore_'  -timeout 180m
=== RUN   TestAccEventDataStore_basic
=== PAUSE TestAccEventDataStore_basic
=== RUN   TestAccEventDataStore_disappears
=== PAUSE TestAccEventDataStore_disappears
=== RUN   TestAccEventDataStore_tags
=== PAUSE TestAccEventDataStore_tags
=== RUN   TestAccEventDataStore_options
=== PAUSE TestAccEventDataStore_options
=== RUN   TestAccEventDataStore_advancedEventSelector
=== PAUSE TestAccEventDataStore_advancedEventSelector
=== CONT  TestAccEventDataStore_basic
=== CONT  TestAccEventDataStore_options
=== CONT  TestAccEventDataStore_advancedEventSelector
=== CONT  TestAccEventDataStore_disappears
=== CONT  TestAccEventDataStore_tags
=== CONT  TestAccEventDataStore_options
    acctest.go:701: this AWS account must be the management account of an AWS Organization
--- SKIP: TestAccEventDataStore_options (1.25s)
--- PASS: TestAccEventDataStore_disappears (16.09s)
--- PASS: TestAccEventDataStore_basic (20.89s)
--- PASS: TestAccEventDataStore_advancedEventSelector (24.49s)
--- PASS: TestAccEventDataStore_tags (41.64s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/cloudtrail	45.575s
```

##### GovCloud

```console
% make testacc TESTS=TestAccEventDataStore_ PKG=cloudtrail
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudtrail/... -v -count 1 -parallel 20 -run='TestAccEventDataStore_'  -timeout 180m
=== RUN   TestAccEventDataStore_basic
=== PAUSE TestAccEventDataStore_basic
=== RUN   TestAccEventDataStore_disappears
=== PAUSE TestAccEventDataStore_disappears
=== RUN   TestAccEventDataStore_tags
=== PAUSE TestAccEventDataStore_tags
=== RUN   TestAccEventDataStore_options
=== PAUSE TestAccEventDataStore_options
=== RUN   TestAccEventDataStore_advancedEventSelector
=== PAUSE TestAccEventDataStore_advancedEventSelector
=== CONT  TestAccEventDataStore_basic
=== CONT  TestAccEventDataStore_options
=== CONT  TestAccEventDataStore_disappears
=== CONT  TestAccEventDataStore_tags
=== CONT  TestAccEventDataStore_advancedEventSelector
=== CONT  TestAccEventDataStore_disappears
    acctest.go:1008: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating CloudTrail Event Data Store (tf-acc-test-8727672389834937490): UnsupportedOperationException: The operation requested is not supported in the region.
        
          with aws_cloudtrail_event_data_store.test,
          on terraform_plugin_test.tf line 2, in resource "aws_cloudtrail_event_data_store" "test":
           2: resource "aws_cloudtrail_event_data_store" "test" {
        
=== CONT  TestAccEventDataStore_basic
    acctest.go:1008: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating CloudTrail Event Data Store (tf-acc-test-3181324507280160273): UnsupportedOperationException: The operation requested is not supported in the region.
        
          with aws_cloudtrail_event_data_store.test,
          on terraform_plugin_test.tf line 2, in resource "aws_cloudtrail_event_data_store" "test":
           2: resource "aws_cloudtrail_event_data_store" "test" {
        
=== CONT  TestAccEventDataStore_advancedEventSelector
    acctest.go:1008: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating CloudTrail Event Data Store (tf-acc-test-6934047121505972719): UnsupportedOperationException: The operation requested is not supported in the region.
        
          with aws_cloudtrail_event_data_store.test,
          on terraform_plugin_test.tf line 2, in resource "aws_cloudtrail_event_data_store" "test":
           2: resource "aws_cloudtrail_event_data_store" "test" {
        
=== CONT  TestAccEventDataStore_options
    acctest.go:1008: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating CloudTrail Event Data Store (tf-acc-test-1237862277570900657): UnsupportedOperationException: The operation requested is not supported in the region.
        
          with aws_cloudtrail_event_data_store.test,
          on terraform_plugin_test.tf line 2, in resource "aws_cloudtrail_event_data_store" "test":
           2: resource "aws_cloudtrail_event_data_store" "test" {
        
=== CONT  TestAccEventDataStore_tags
    acctest.go:1008: skipping test for aws-us-gov/us-gov-west-1: Error running apply: exit status 1
        
        Error: error creating CloudTrail Event Data Store (tf-acc-test-6132664734702259425): UnsupportedOperationException: The operation requested is not supported in the region.
        
          with aws_cloudtrail_event_data_store.test,
          on terraform_plugin_test.tf line 2, in resource "aws_cloudtrail_event_data_store" "test":
           2: resource "aws_cloudtrail_event_data_store" "test" {
        
--- SKIP: TestAccEventDataStore_disappears (9.13s)
--- SKIP: TestAccEventDataStore_advancedEventSelector (9.19s)
--- SKIP: TestAccEventDataStore_options (9.21s)
--- SKIP: TestAccEventDataStore_basic (9.21s)
--- SKIP: TestAccEventDataStore_tags (9.22s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/cloudtrail	12.617s
```

**Comment by ewbankkit** [comment]
@bschaatsbergen Thanks for the contribution :tada: :clap:.
Overall the contribution was great.
I made a few mainly cosmetic changes and fixed some bugs in the resource Update functionality that additional tests uncovered.

**Comment by bschaatsbergen** [comment]
@ewbankkit. that's great to hear and for sure more contributions will follow. Thanks for adding the touch-ups!

**Pull request merged**